src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java

```diff
@@ -1860,21 +1860,37 @@ else if ("test".equals(configName)) {
     }
 
     private static void validateChapterWiseTesting(Iterator<Node> itrChecks,
-          Iterator<Node> itrConfigs, Set<String> styleChecks, String styleName, String ruleName) {
+          Iterator<Node> itrSample, Set<String> styleChecks, String styleName, String ruleName) {
         boolean hasChecks = false;
+        final Set<String> checkLinks = new HashSet<>();
 
         while (itrChecks.hasNext()) {
             final Node module = itrChecks.next();
             final String moduleName = module.getTextContent().trim();
             final String href = module.getAttributes().getNamedItem("href").getTextContent();
             final boolean moduleIsCheck = href.startsWith("checks/");
 
+            final String partialConfigUrl = "https://github.com/search?q="
+                    + "path%3Asrc%2Fmain%2Fresources%20path%3A**%2F" + styleName;
+
             if (!moduleIsCheck) {
+                if (href.startsWith(partialConfigUrl)) {
+                    assertWithMessage("google_style.xml rule '" + ruleName + "' module '"
+                            + moduleName + "' has too many config links").fail();
+                }
                 continue;
             }
 
             hasChecks = true;
 
+            assertWithMessage("The module '" + moduleName + "' in the rule '" + ruleName
+                    + "' of the style guide '" + styleName
+                    + "_style.xml' should not appear more than once in the section.")
```

Comment:
This message is not what this assertion is checking. You are just checking that a link doesn't appear twice.

If you want to verify that a module doesn't appear twice, then I believe you would attach this to the `styleChecks.remove(moduleName);` and look for a remove to fail (it wasn't removed since it doesn't exist in the list).

Reply:
`styleChecks` contains all the modules used in style guide, it does not contains the modules used in the particular section:

https://github.com/checkstyle/checkstyle/blob/c90ad6517035f97a3187c0037700acf311b67477/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java#L1615-L1619

I used it something like:

```java
            assertWithMessage("The module '" + moduleName + "' in the rule '" + ruleName
                    + "' of the style guide '" + styleName
                    + "_style.xml' should not appear more than once in the section.")
                    .that(styleChecks.remove(moduleName))
                    .isTrue();
```

and it was giving duplicate module usage for LineLength config of section [3.3.2](https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/c90ad65_2024115000/google_style.html#a3.3.2) because it was used in earlier section [3.2](https://checkstyle-diff-reports.s3.us-east-2.amazonaws.com/c90ad65_2024115000/google_style.html#a3.2)

Reply:
diff --git a/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java b/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java
index 741ed0a..799dee0 100644
--- a/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java
@@ -1884,9 +1884,8 @@ public class XdocsPagesTest {
 
             hasChecks = true;
 
-            assertWithMessage("The module '" + moduleName + "' in the rule '" + ruleName
-                    + "' of the style guide '" + styleName
-                    + "_style.xml' should not appear more than once in the section.")
+            assertWithMessage("Link '" + href + "' should not be present twice in "
+                    + styleName + "_style.xml")
                     .that(checkLinks)
                     .doesNotContain(href);
 
diff --git a/src/xdocs/google_style.xml b/src/xdocs/google_style.xml
index b1d12a6..2caba26 100644
--- a/src/xdocs/google_style.xml
+++ b/src/xdocs/google_style.xml
@@ -2378,8 +2378,8 @@
                         src="images/ok_green.png"
                         alt="" />
                   </span>
-                  <a href="checks/javadoc/missingjavadocmethod.html#MissingJavadocMethod">MissingJavadocMethod</a>
-                  (<a href="https://github.com/search?q=path%3Asrc%2Fmain%2Fresources%20path%3A**%2Fgoogle_checks.xml+repo%3Acheckstyle%2Fcheckstyle+MissingJavadocMethod">
+                  <a href="checks/javadoc/javadocmethod.html#JavadocMethod">JavadocMethod</a>
+                  (<a href="https://github.com/search?q=path%3Asrc%2Fmain%2Fresources%20path%3A**%2Fgoogle_checks.xml+repo%3Acheckstyle%2Fcheckstyle+JavadocMethod">
                   config</a>)
                   <br />
                   <br />

```

result in:
```
Link 'checks/javadoc/javadocmethod.html#JavadocMethod' should not be present twice in google_style.xml
expected not to contain: checks/javadoc/javadocmethod.html#JavadocMethod
but was                : [checks/javadoc/javadocmethod.html#JavadocMethod]

	at com.puppycrawl.tools.checkstyle.internal.XdocsPagesTest.validateChapterWiseTesting(XdocsPagesTest.java:1890)
	at com.puppycrawl.tools.checkstyle.internal.XdocsPagesTest.validateStyleModules(XdocsPagesTest.java:1782)
```

Reply:
@rnveach , done.

Reply:
We are saying we can have duplicate modules but not duplicate links? It doesn't make sense to count links.

````
-                  <a href="checks/javadoc/missingjavadocmethod.html#MissingJavadocMethod">MissingJavadocMethod</a>
-                  (<a href="https://github.com/search?q=path%3Asrc%2Fmain%2Fresources%20path%3A**%2Fgoogle_checks.xml+repo%3Acheckstyle%2Fcheckstyle+MissingJavadocMethod">
+                  <a href="checks/javadoc/javadocmethod.html#JavadocMethod">JavadocMethod</a>
+                  (<a href="https://github.com/search?q=path%3Asrc%2Fmain%2Fresources%20path%3A**%2Fgoogle_checks.xml+repo%3Acheckstyle%2Fcheckstyle+JavadocMethod">
````

So how did we miss this and nothing in our automated testing caught it?

Reply:
Let's remove this validation at all .
We can have same module even with same id in multiple sections.

Reply:
The same module is not needed in a single section. Our link to config already covers multiple instances in the config. We should not see multiple modules in a single section.

Reply:
I agree with Richard, we should not have duplicate modules in single section that's why I added this new test. We should not remove this

Reply:
should we change test to validate Module Name instead of module link? I thought checking module link would be more accurate than checking the module name

Reply:
> The same module is not needed in a single section.

yes, but it is so easy to catch during code review. We can definitely make such validation, but value of it will not be big.
I am to make it only it is simple. If it requires more refactoring, lets postpone it for future.

Reply:
There are more PR standing to update table content with new details lets merge this PR sooner.

Reply:
Since @Zopsss asked me here, I will try and continue with this PR.

@romani Please let me know if you are going to ignore my review and do what you want anyways. It will save us both the time of going back and forth on this and allow you to do what you are going to do anwyays.

https://github.com/checkstyle/checkstyle/issues/15643#issuecomment-2336777333

Reply:
I am not ignoring. I just suggesting to not go deep too much, and we need update more, do it few PRs, to unblock structural changes of table that this PR is doing.
We can start with smaller set of test for structure valudation and improve in next PR that will be a way smaller.

Reply:
Is that a yes or no?

Reply:
>Please let me know if you are going to ignore my review

No

Reply:
@Zopsss We should check both the module name and the link. It doesn't make sense to check for duplicate links. Validate for duplicate module names in the same section. Then once the check name is validated, validate the link is what it should be for the module.

Reply:
done

https://github.com/checkstyle/checkstyle/blob/664a71df19233e1e6444dc07ea9e01e564eab158/src/test/java/com/puppycrawl/tools/checkstyle/internal/XdocsPagesTest.java#L1887-L1898

first duplicate module name is validated then the link is validated. We are only checking that it should not end with `Check`, we're cannot validate anything else than this.